src/Core/Unix/FreeBSD/CoreFreeBSD.cpp

```diff
@@ -185,17 +185,58 @@ namespace VeraCrypt
 
 	void CoreFreeBSD::MountFilesystem (const DevicePath &devicePath, const DirectoryPath &mountPoint, const string &filesystemType, bool readOnly, const string &systemMountOptions) const
 	{
+		std::string chosenFilesystem = "msdos";
+		std::string modifiedMountOptions = systemMountOptions;
+
+		if (filesystemType.empty() && modifiedMountOptions.find("mountprog") == string::npos) {
+			// No filesystem type specified through CLI, attempt to identify with blkid
+			// as mount is unable to probe filesystem type on BSD
+			// Make sure we don't override user defined mountprog
+			std::vector<char> buffer(128,0);
+			std::string cmd = "blkid -o value -s TYPE " + static_cast<std::string>(devicePath) + " 2>/dev/null";
+			std::string result;
+
+			FILE* pipe = popen(cmd.c_str(), "r");
+			if (pipe) {
+				while (!feof(pipe)) {
+					if (fgets(buffer.data(), 128, pipe) != nullptr)
+						result += buffer.data();
+				}
+				fflush(pipe);
+				pclose(pipe);
+				pipe = nullptr;
+			}
+
+			if (result.find("ext") == 0 || StringConverter::ToLower(filesystemType).find("ext") == 0) {
+				chosenFilesystem = "ext2fs";
+			}
+			else if (result.find("exfat") == 0 || StringConverter::ToLower(filesystemType) == "exfat") {
+				chosenFilesystem = "exfat";
+				modifiedMountOptions += string(!systemMountOptions.empty() ? "," : "")
+							+ "mountprog=/usr/local/sbin/mount.exfat";
+			}
+			else if (result.find("ntfs") == 0 || StringConverter::ToLower(filesystemType) == "ntfs") {
+				chosenFilesystem = "ntfs";
+				modifiedMountOptions += string(!systemMountOptions.empty() ? "," : "")
+							+ "mountprog=/usr/local/bin/ntfs-3g";
+			}
+			else if (!filesystemType.empty()) {
+				// Filesystem is specified but is none of the above, then supply as is
+				chosenFilesystem = filesystemType;
+			}
+		} else
+			chosenFilesystem = filesystemType;
+
 		try
 		{
-			// Try to mount FAT by default as mount is unable to probe filesystem type on BSD
-			CoreUnix::MountFilesystem (devicePath, mountPoint, filesystemType.empty() ? "msdos" : filesystemType, readOnly, systemMountOptions);
+			CoreUnix::MountFilesystem (devicePath, mountPoint, chosenFilesystem, readOnly, modifiedMountOptions);
 		}
 		catch (ExecutedProcessFailed&)
 		{
 			if (!filesystemType.empty())
 				throw;
 
-			CoreUnix::MountFilesystem (devicePath, mountPoint, filesystemType, readOnly, systemMountOptions);
+			CoreUnix::MountFilesystem (devicePath, mountPoint, filesystemType, readOnly, modifiedMountOptions);
```

Comment:
@Jertzukka :
Is there a specific reason why `modifiedMountOptions `is retained in the catch block instead of reverting to `systemMountOptions`? `modifiedMountOptions `is directly linked to `chosenFilesystem `so using it with `filesystemType  `doesn't look right

Reply:
Only paths where `modifiedMountOptions` changes is if:

- blkid supplies exfat or ntfs result, in this case using `systemMountOptions` which omits our specified mountprog will never successfully mount.
- User specifies `--filesystem=exfat` or `--filesystem=ntfs` through CLI, let's say the volume is actually UFS. First mount fails, we throw because `if (!filesystemType.empty()) throw;` and we never get to the second mount prompt.

I remember thinking this through and coming to a conclusion that in practice using either `systemMountOptions` and `modifiedMountOptions` here has no difference, as the only edge case where it matters would be where blkid misidentifies a volume as ntfs/exfat where it was actually something like UFS. But I agree, it is probably clearer for anyone reading the code not having to do 10 steps of logic to see why a variable that refers to something `modified` is used here in a fallback path.